docs/compiler-notes.rst

```diff
@@ -0,0 +1,219 @@
+- `with`
+- `while`
+- `global`
+- `nonlocal`
+- `lambda` functions
+- bitwise shifting operators
+  - The use of bitwise shiftiing operators is not supported, instead `bitewise_*` functions exist.
+- shifting operators not supported. Use `shift` insetad.
+- floor division operator `//` is not supported. Use `floor` and `ceil` instead, to make rounding
+  operations clear.
+
+Statements
+~~~~~~~~~~
+
+*Simple Statements*
+
+**Airthmetic Conversions**
+
+Vyper enforces airthmetic conversions, to be explicit vs. python which uses implicit type conversion.
+This was chosen specifically, to make the conversion steps clearer. The expression `1 * 1.0` will
+be disallowed, instead `convert(1, decimal) * 1.0`.
+
+**Identifiers**
+
+Vyper ensures variables names can only consist of alphanumeric characters, with `_`. The specific
+regex tested against is `^[_a-zA-Z][a-zA-Z0-9_]*$`.
+
+**Literals**
+
+Vyper has support for integer, decimal, string and byte python literals.
+- If to Literals are found in an expression they are constant folded.
+  - `10*10` will be converted to `100`
+  - However `10*10*a` will not be converted.
+- See above artihmetic onversion rules.
+- There are limitations to byte & string literals based on their usage.
+  - e.g. A string literal of size 100 can't be assigned or return to an annoted size of 32.
+- To distinguish between `int128` and `uint256` literals Vyper uses the following ruleset.
+  - If literal int value is above MAXNUM (2**127 - 1), literal is a uint256.
+  - If literal int value is under MINNUM (-2**127), literal is invalid.
+  - If assignment of literal to a known type (could be uint256 or int128), use appropriate type.
+  - Consequently slightly confusing error message of `Cannot assign uint256 to int128` does occur,
+    but seemed to be a reasonable trade off (don't have to convert(xxxx, uint256) on assingment).
+
+**Comparisons / Boolean Operations*
+
+- Comparisons
+  - As long as all members of are of the same type -comparisons are supported.
+- Boolean `and`, `or` operations are supported.
+    - Currently not short-circuited (VIP pending)
+
+** Membership **
+
+- Array membership e.g. `assert a in self.owners` is supported.
+
+** del Delete Statement **
+
+Vyper does not support `del ..` to clear values from storage use `clear()` instead.
+
+** Import Statements **
+
+Import statements are supported for use with import interfaces, and do not import functions.
+
+** Assert Statements **
+
+`assert` is supported and uses the `REVERT` opcode. To `INVALID` opcode, use assert <expre>, UNREACHABLE
+
+** Raise Statements **
+
+`raise` is supported, uses the `REVERT` opcode to throw an exception (with reason string smaller 
+than 32 bytes)
+
+** Return Statements **
+
+- `return` statements are supported.
+- There is a checker in place to ensure return statements are not missing in branches.
+
+*Compound Statements*
+
+**Functions**
+
+- Vyper support default arguments for both public and private functions.
```

Comment:
What happens with default arguments in private functions? Do we have test cases for this?